kolibri/plugins/coach/assets/src/composables/useFetch.js

```diff
@@ -0,0 +1,199 @@
+ * * You can use `lodash.get`-compatible keys to access nested objects (e.g., "data.user.details").
+ *
+ * Example:
+ * ```js
+ * // Suppose the response object looks like this:
+ * const response = {
+ *   payload: [...],
+ *   more: { page: 2 }
+ * };
+ *
+ * // By specifying `moreKey`, you tell the composable where to find the more object
+ * // to fetch additional data:
+ * useFetch({ moreKey: "more" });
+ * ```
+ *
+ *
+ * @param {(more, ...args) => Promise<any>} [options.fetchMoreMethod] Function to fetch more data.
+ * * This function receives a "more" object as its first argument. This "more" object is specified
+ *   by the `moreKey` param.
+ * * Should return a Promise resolving to the additional data.
+ * * FetchMore just works if the fetched data is an array
+ *
+ * Example:
+ *
+ * ```js
+ * const fetchMoreMethod = (more) => ContentNodeResource.fetchBookmarks({
+ *  params: more
+ * })
+ * ```
+ *
+ *
+ * @param {Object.<string, string>} [options.additionalDataKeys] An object that maps additional
+ * data keys for the response object.
+ * * In the `{ key: value }` pair:
+ *   - The `key` will be used as the property name in the returned `additionalData` object.
+ *   - The `value` specifies the key in the response object from which the data will be retrieved.
+ *
+ * Example:
+ * ```js
+ * const additionalDataKeys = {
+ *   userId: "user_id", // The `userId` property in `additionalData` will map to `response.user_id`
+ *   userName: "name" // The `userName` property in `additionalData` will map to `response.name`
+ * };
+ *
+ * const { additionalData } = useFetch({ additionalDataKeys });
+ * console.log(additionalData.userId);  // Outputs the value of `response.user_id`
+ * console.log(additionalData.userName); // Outputs the value of `response.name`
+ * ```
+ *
+ *
+ * @typedef {Object} FetchObject
+ * @property {any} data The main fetched data.
+ * @property {Object} error Error object if a fetch data failed.
+ * @property {boolean} loading Data loading state. This loading doesnt reflect the loading when
+ *   fetching more data. refer to moreState for that.
+ * @property {string} moreState State of the fetch more data, it could be LOADING, HAS_MORE,
+ *   NO_MORE or ERROR.
+ * @property {Object<string, any>} additionalData Extra data specified by `additionalDataKeys`.
+ * @property {(...args) => Promise<void>} fetchData A method to manually trigger the main fetch.
+ * @property {(...args) => Promise<void>} fetchMore A method to manually trigger fetch more data.
+ *
+ * @returns {FetchObject} An object with properties and methods for managing the fetch process.
+ */
+export default function useFetch(options) {
+  const { fetchMethod, fetchMoreMethod, dataKey, moreKey, additionalDataKeys } = options || {};
+
+  const loading = ref(false);
+  const data = ref(null);
+  const error = ref(null);
+  const more = ref(null);
+  const loadingMore = ref(false);
+  const additionalData = ref(null);
+
+  const moreState = computed(() => {
```

Comment:
I am not sure this state gives us anything over just exporting `more` and `loadingMore` as simple booleans.

Reply:
Yes, we can export both. It was just to pass it directly to the vue component using the states of the `ViewMoreButtonStates`, but I dont have a strong preference. And perhaps returning both booleans can be a little more absctract.

Reply:
Updated!